Autopilot/AttitudeManager/OrientationControl.c

```diff
@@ -1,146 +1,140 @@
+const static float init_kp[PID_CHANNELS] = {1, 1, 1, 1, 1, 1, 1, 1};
+const static float init_ki[PID_CHANNELS] = {0, 0, 0, 0, 0, 0, 0, 0};
+const static float init_kd[PID_CHANNELS] = {0, 0, 0, 0, 0, 0, 0, 0};
+
+/* Generic PID functions. Can be used to PID other things (flaps, etc) */
+
+// To be called to initialize a new PID channel
+void initPID(PID_val* pid, float Kp, float Ki, float Kd, int16_t imax) {
+    pid->Kp = Kp;
+    pid->Ki = Ki;
+    pid->Kd = Kd;
+    pid->imax = imax;
+    pid->integral = 0;
+    pid->lastTime = 0;
+    pid->lastErr = 0;
+    pid->lastDer = 0;
 }
 
-int controlSignalHeading(int setpoint, int output) { // function to find output based on gyro acceleration and PWM input
-    //Take into account Heading overflow (330 degrees and 30 degrees is a 60 degree difference)
-    if (setpoint - output > 180){
-        output += 360;
-    }
-    else if (setpoint - output < -180){
-        output -= 360;
-    }
+// PID loop function. error is (setpointValue - currentValue)
+float PIDcontrol(PID_val* pid, float error, float scale) {
+    float output = 0;
+        
+    uint32_t now = getTime();
+    uint16_t delta_msec = (now - pid->lastTime);
     
-    int error = setpoint - output;
-    if (integralFreeze == 0){
-        sum_gain[HEADING] += (float)error;
+    // check if we've gone too long without updating (keeps the I and D from freaking out)
+    if (delta_msec > PID_RESET_TIME || pid->lastTime == 0) {
+        delta_msec = 0;
+        pid->integral = 0;
+        pid->lastErr = error;
     }
     
-    //Derivative Calculations
-    int dValue = error - lastError[HEADING];
-    lastError[HEADING] = error;
+    pid->lastTime = now;
 
-    int controlSignal = (int)(HEADING_ROLL_SCALE_FACTOR * ((dValue * kd_gain[HEADING]) + (error * kp_gain[HEADING]) + (sum_gain[HEADING] * ki_gain[HEADING])));
-    return controlSignal;
-}
-int controlSignalAngles(float setpoint, float output, unsigned char type, float SERVO_SCALE_FACTOR_ANGLES) { // function to find output based on gyro acceleration and PWM input
+    output += pid->Kp * error; // Proportional control
+    
+    if (delta_msec > 0) { // only compute time-sensitive control if time has elapsed
+        float dTime = delta_msec / 1000.0f; // elapsed time in seconds
+        
+        if (fabsf(pid->Ki) > 0) { // Integral control
+            pid->integral += (pid->Ki * error) * dTime;
+            
+            if (pid->integral < -pid->imax) { // ensure integral stays manageable
+                pid->integral = -pid->imax;
+            } else if (pid->integral > pid->imax) {
+                pid->integral = pid->imax;
+            }
+            output += pid->integral;
+        }
 
-    float error = setpoint - output;
-    if (integralFreeze == 0){
-        sum_gain[type] += error;
+        if (fabsf(pid->Kd) > 0) { // Derivative control
+            float derivative = (error - pid->lastErr) / dTime;
```

Comment:
Pretty much, yeah. With milliseconds, You'd need really big KD values and really small KI ones.